lib/openssl/buffering.rb

```diff
@@ -348,15 +348,19 @@ def do_write(s)
     @wbuffer << s
     @wbuffer.force_encoding(Encoding::BINARY)
     @sync ||= false
+    @offset ||= 0
+
     if @sync or @wbuffer.size > BLOCK_SIZE
-      until @wbuffer.empty?
+      while @offset < @wbuffer.size
         begin
-          nwrote = syswrite(@wbuffer)
+          @offset += syswrite(@wbuffer[@offset, BLOCK_SIZE])
         rescue Errno::EAGAIN
           retry
         end
-        @wbuffer[0, nwrote] = ""
       end
+
+      @wbuffer.clear
```

Comment:
I would advise to use a double buffer approach here. Or at least we need to consider that `@wbuffer` could in theory be modified while writing. Not sure if this is a theoretical problem or a real problem.

Reply:
Hi @ioquatix ,

Thank you for your advise.
Would you like to change it as below?

```ruby
  def do_write(s)
    @wbuffer = Buffer.new unless defined? @wbuffer
    @wbuffer << s
    @wbuffer.force_encoding(Encoding::BINARY)
    @sync ||= false

    if @sync or @wbuffer.size > BLOCK_SIZE
      copied = @wbuffer.clone
      @wbuffer[0, copied.size] = ""

      offset = 0
      while offset < copied.size
        begin
          offset += syswrite(copied[offset, BLOCK_SIZE])
        rescue Errno::EAGAIN
          retry
        end
      end
    end
  end

# Total allocated: 40503078 bytes (1648 objects)
# Total retained:  11535026 bytes (6 objects)
```

But it's not thread-safe code originally., such as codes below.

```ruby
@wbuffer = Buffer.new unless defined? @wbuffer
...
@sync ||= false
```

So I don't think we have to care about it.